File: leetcode-algorithms/692. Top K Frequent Words/Solution.cpp
// Given a non-empty list of words, return the k most frequent elements.
//
// Your answer should be sorted by frequency from highest to lowest. If two words have the same frequency, then the word with the lower alphabetical order comes first.
//
// Example 1:
// Input: ["i", "love", "leetcode", "i", "love", "coding"], k = 2
// Output: ["i", "love"]
// Explanation: "i" and "love" are the two most frequent words.
//     Note that "i" comes before "love" due to a lower alphabetical order.
// Example 2:
// Input: ["the", "day", "is", "sunny", "the", "the", "the", "sunny", "is", "is"], k = 4
// Output: ["the", "is", "sunny", "day"]
// Explanation: "the", "is", "sunny" and "day" are the four most frequent words,
//     with the number of occurrence being 4, 3, 2 and 1 respectively.
// Note:
// You may assume k is always valid, 1 ≤ k ≤ number of unique elements.
// Input words contain only lowercase letters.
// Follow up:
// Try to solve it in O(n log k) time and O(n) extra space.
struct data {
    data(string val_, int count_) { val = val_; count = count_; }
    string val;
    int count;
};

class cmp {
public:
    bool operator()(data& a, data& b) {
        if (a.count == b.count) {
            return a.val < b.val;
        } else return a.count > b.count;
    }
};

class Solution {
public:
    vector<string> topKFrequent(vector<string>& words, int k) {
        map<string, int> smap; // word->count;

        for (auto word: words) {
            smap[word]++;;
        }

        priority_queue<data, vector<data>, cmp> queue;

        for (auto item : smap) {
            queue.push(data(item.first, item.second));
            if (queue.size() > k) queue.pop();
        }

        vector<string> rslt;
        while (!queue.empty()) {
            rslt.insert(rslt.begin(), queue.top().val);
            queue.pop();
        }
        return rslt;
    }
};

File: leetcode-algorithms/692. Top K Frequent Words/Solution.java
class itemaa {
	public itemaa(String str, int count) {
		this.str = str;
		this.count = count;
	}

	public String str;
	public int count;
}
class Solution {
    public List<String> topKFrequent(String[] words, int k) {
        Map<String, Integer> vmap = new HashMap<>();
        for (String word : words) {
        	vmap.put(word, vmap.getOrDefault(word, 0) + 1);
        }

        List<itemaa> va =   vmap.keySet().stream().map(key -> new itemaa(key, vmap.get(key))).collect(Collectors.toList());

        PriorityQueue<itemaa> que = new PriorityQueue<>((a, b) -> {
        	if (a.count != b.count) return a.count - b.count;
        	else return b.str.compareTo(a.str);
        });

        for (itemaa item : va) {
        	if (que.size() < k) que.add(item);
        	else if (((que.peek().count == item.count) && (item.str.compareTo(que.peek().str) < 0)) || (que.peek().count < item.count)) {
        		que.poll();
        		que.add(item);
        	}
        }

        List<String> rslt = new ArrayList<>();
        while (!que.isEmpty()) {
        	rslt.add(0, que.poll().str);
        }

        return rslt;
    }
}
